src/ucp/rndv/proto_rndv.c

```diff
@@ -365,41 +348,147 @@ ucp_proto_rndv_ctrl_init(const ucp_proto_rndv_ctrl_init_params_t *params,
+    ucs_log_indent(+1);
+    ucs_array_for_each(remote_proto, &remote_proto_init->protocols) {
+        if (ucp_proto_id_field(remote_proto->proto_id, flags) &
+            UCP_PROTO_FLAG_INVALID) {
+            continue;
+        }
+
+        ucs_assertv(remote_proto->caps.num_ranges > 0,
+                    "remote_proto->caps.num_ranges=%u",
+                    remote_proto->caps.num_ranges);
+
+        proto_name = ucp_proto_id_field(remote_proto->proto_id, name);
+        caps       = &remote_proto->caps;
+        max_length = caps->ranges[caps->num_ranges - 1].max_length;
+        min_length = caps->min_length;
+        ucs_trace("%s proto variant: min_length=%zu max_length=%zu "
+                  "cfg_thresh=%zu cfg_priority=%u", proto_name, 
+                  min_length, max_length, remote_proto->cfg_thresh,
+                  remote_proto->cfg_priority);
+
+        max_length = ucs_min(max_length, params->super.max_length);
+        min_length = ucs_max(min_length, params->super.min_length);
+        if (max_length < min_length) {
+            ucs_trace("no available ranges for %s", proto_name);
+            continue;
+        }
+
+        remote_priv = &ucs_array_elem(&remote_proto_init->priv_buf,
+                                      remote_proto->priv_offset);
+        ucp_proto_rndv_set_variant_config(init_params, remote_proto,
+                                          &remote_select_param, remote_priv,
+                                          &rpriv->remote_proto_config);
+
+        /* Create separate CTRL message perf node for each variant */
+        status = ucp_proto_rndv_ctrl_range_init(params, rpriv, &ctrl_perf);
+        if (status != UCS_OK) {
+            ucs_trace("failed to init perf data for %s", proto_name);
+            continue;
+        }
+
+        ucp_proto_common_init_base_caps(&params->super, &result_caps,
+                                        min_length);
+
+        status = ucp_proto_rndv_ctrl_init_parallel_stages(
+                params, &ctrl_perf, caps->ranges, &result_caps, min_length,
+                max_length);
+        ucp_proto_perf_node_deref(&ctrl_perf.node);
+        if (status != UCS_OK) {
+            ucs_trace("failed to calculate parallel stages for %s", proto_name);
+            continue;
+        }
+
+        cfg_thresh   = params->super.cfg_thresh;
+        cfg_priority = params->super.cfg_priority;
+        if ((remote_proto->cfg_thresh != UCS_MEMUNITS_INF) &&
+            (remote_proto->cfg_thresh != UCS_MEMUNITS_AUTO)) {
+            /* Consider remote priority and threshold only if RNDV_SCHEME or
+             * BCOPY/ZCOPY thresh are set to force these settings */
+            cfg_priority = remote_proto->cfg_priority;
+            cfg_thresh   = ucs_max(cfg_thresh, remote_proto->cfg_thresh);
+        }
 
-    ucp_proto_common_add_proto(&params->super, &caps, &priv, sizeof(priv));
+        ucs_assertv(params->super.cfg_priority <= remote_proto->cfg_priority,
```

Comment:
why?

Reply:
It is a design requirement, since if variant cfg_thresh not inf or auto, then we took cfg_priority from variant to force it among others. That's why it is required that it should be higher than params->cfg_priority

Reply:
maybe better to add comment about it here.